File: python/problems/leetcode/0095_unique_binary_search_trees_ii/solution.py
"""LeetCode 95: Unique Binary Search Trees II."""

from __future__ import annotations

from typing import Optional


class TreeNode:
    def __init__(
        self,
        val: int = 0,
        left: Optional["TreeNode"] = None,
        right: Optional["TreeNode"] = None,
    ) -> None:
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def generateTrees(self, n: int) -> list[Optional[TreeNode]]:
        if n == 0:
            return []

        def build(l: int, r: int) -> list[Optional[TreeNode]]:
            if l > r:
                return [None]
            out: list[Optional[TreeNode]] = []
            for i in range(l, r + 1):
                for lt in build(l, i - 1):
                    for rt in build(i + 1, r):
                        out.append(TreeNode(i, lt, rt))
            return out

        return build(1, n)


def inorder(root: Optional[TreeNode]) -> list[int]:
    out: list[int] = []

    def dfs(n: Optional[TreeNode]) -> None:
        if n is None:
            return
        dfs(n.left)
        out.append(n.val)
        dfs(n.right)

    dfs(root)
    return out


if __name__ == "__main__":
    trees = Solution().generateTrees(3)
    assert len(trees) == 5
    for t in trees:
        seq = inorder(t)
        assert seq == sorted(seq)
    assert Solution().generateTrees(0) == []
    print("0095 OK")

File: cpp/problems/leetcode/0095_unique_binary_search_trees_ii/solution.cpp
// LeetCode 95
#include <bits/stdc++.h>
#include <cassert>
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int v) : val(v), left(nullptr), right(nullptr) {}
    TreeNode(int v, TreeNode* l, TreeNode* r) : val(v), left(l), right(r) {}
};

class Solution {
public:
    vector<TreeNode*> generateTrees(int n) {
        if (n == 0) return {};
        return build(1, n);
    }

private:
    vector<TreeNode*> build(int l, int r) {
        if (l > r) return {nullptr};
        vector<TreeNode*> out;
        for (int i = l; i <= r; ++i)
            for (auto* lt : build(l, i - 1))
                for (auto* rt : build(i + 1, r))
                    out.push_back(new TreeNode(i, lt, rt));
        return out;
    }
};

static void inorderTraverse(TreeNode* n, vector<int>& out) {
    if (!n) return;
    inorderTraverse(n->left, out);
    out.push_back(n->val);
    inorderTraverse(n->right, out);
}

int main() {
    auto trees = Solution().generateTrees(3);
    assert((int)trees.size() == 5);
    for (auto* t : trees) {
        vector<int> seq;
        inorderTraverse(t, seq);
        auto sorted_seq = seq;
        sort(sorted_seq.begin(), sorted_seq.end());
        assert(seq == sorted_seq);
    }
    assert(Solution().generateTrees(0).empty());
    cout << "0095 OK" << endl;
    return 0;
}
